pypeit/images/buildimage.py

```diff
@@ -0,0 +1,192 @@
+""" Object to hold + process a single image"""
+
+import inspect
+
+import os
+import numpy as np
+
+
+from pypeit import msgs
+
+from pypeit.core import coadd2d
+from pypeit.par import pypeitpar
+from pypeit import utils
+
+from pypeit.images import pypeitimage
+from pypeit.images import processrawimage
+from pypeit.images import rawimage
+from pypeit.images import maskimage
+
+from IPython import embed
+
+
+class BuildImage(object):
+    """
+    Class to generate an image from one or more files (and other pieces).
+
+    The core processing steps are handled by ProcessRawImage
+    This object is mainly for combining multiple images
+
+    Args:
+        spectrograph (:class:`pypeit.spectrographs.spectrograph.Spectrograph`):
+            Spectrograph used to take the data.
+        det (:obj:`int`, optional):
+            The 1-indexed detector number to process.
+        par (:class:`pypeit.par.pypeitpar.ProcessImagesPar`):
+            Parameters that dictate the processing of the images.  See
+            :class:`pypeit.par.pypeitpar.ProcessImagesPar` for the
+            defaults.
+
+    """
+    def __init__(self, spectrograph, det, par, files):
+
+        # Required parameters
+        self.spectrograph = spectrograph
+        self.det = det
+        if not isinstance(par, pypeitpar.ProcessImagesPar):
+            msgs.error('Provided ParSet for must be type ProcessImagesPar.')
+        self.par = par  # This musts be named this way as it is frequently a child
+        self.files = files
+
+    def process_one(self, filename, process_steps, bias, pixel_flat, illum_flat=None, bpm=None):
+        """
+        Process a single image
+
+        Args:
+            filename (str):
+                File to process
+            process_steps (list):
+                List of processing steps
+            bias (np.ndarray or None):
+                Bias image
+            pixel_flat (np.ndarray):
+                Flat image
+            illum_flat (np.ndarray, optional):
+                Illumination image
+            bpm (np.ndarray, optional):
+                Bad pixel mask
+
+        Returns:
+            :class:`pypeit.images.pypeitimage.PypeItImage`:
+
+        """
+        # Load raw image
+        rawImage = rawimage.RawImage(filename, self.spectrograph, self.det)
+
+        # Process
+        processrawImage = processrawimage.ProcessRawImage(rawImage, self.par, bpm=bpm)
+        processedImage = processrawImage.process(process_steps, bias=bias, pixel_flat=pixel_flat,
+                                                 illum_flat=illum_flat)
+
+        # Return
+        return processedImage
+
+    def run(self, process_steps, bias, pixel_flat=None, illum_flat=None,
```

Comment:
Why is pixel_flat a required argument above, but not a required argument here? In my opinion, only filename should be required. Process steps and bias should be optional. Process_steps should default to just trimming, Oscan_subtraction and gain multiplication. 

I've been begging for the equivalent of 

IDL> long_proc, filename, imag, ivar 

for quite some time now, but never got my wish. Can you make that a reality please in this PR. We can debate what should be optional vs or required, but I think there really needs to be a one line utility function for this class.

Reply:
None of the calibration images use a pixel_flat.  Have made it optional above now.

As for your IDL request, I have a hard time parsing such code, but I believe this does the trick:

```
In [1]: from pypeit.spectrographs import util                                  
In [2]: shaneb = util.load_spectrograph('shane_kast_blue')                     
In [3]: from pypeit.images.processrawimage import process_raw_for_jfh          
In [4]: pypeitImage = process_raw_for_jfh('b13.fits.gz', shaneb)               
[WARNING] :: processrawimage.py 397 process_raw_for_jfh() - Using the Processinng parameters from scienceframe
```

I look forward to hearing how much easier this makes your coding.

Reply:
Thanks for adding this. Even better would be a "header2spectrograph" method, that does not require the spectrograph.

Reply:
I'll let you code that up for the 10+ spectrographs
and their unique headers.  Probably won't be too hard.